internal/cmd/controller/reconciler/cronduration_job.go

```diff
@@ -0,0 +1,323 @@
+		return 0, err
+	}
+
+	return time.Duration(nextFireTime - now), nil
+}
+
+// checkScheduleAndDuration verifies that the given schedule start time and the duration are feasible.
+// If the duration is longer than 2 consecutive triggers it is considered as non valid.
+func checkScheduleAndDuration(schedule *fleet.Schedule, location *time.Location) error { //nolint:revive
+	trigger, err := quartz.NewCronTriggerWithLoc(schedule.Spec.Schedule, location)
+	if err != nil {
+		return err
+	}
+	now := quartz.NowNano()
+	firstFireTime, err := trigger.NextFireTime(now)
+	if err != nil {
+		return err
+	}
+	firstFireTimeSecs := firstFireTime / 1_000_000_000
+
+	secondFireTime, err := trigger.NextFireTime(firstFireTime)
+	if err != nil {
+		return err
+	}
+	secondFireTimeSecs := secondFireTime / 1_000_000_000
+
+	if int64(schedule.Spec.Duration.Seconds()) >= (secondFireTimeSecs - firstFireTimeSecs) {
+		// we also consider an error when duration is equal to the next time
+		// the job should trigger because we could incur race conditions.
+		return fmt.Errorf("duration is too long and overlaps with the next execution time")
+	}
+
+	return nil
+}
+
+func (c *CronDurationJob) scheduleStopJob() error {
+	return c.Scheduler.ScheduleJob(
+		quartz.NewJobDetailWithOptions(
+			c,
+			c.key,
+			&quartz.JobDetailOptions{
+				Replace: true,
+			},
+		),
+		quartz.NewRunOnceTrigger(c.Schedule.Spec.Duration.Duration),
+	)
+}
+
+func (c *CronDurationJob) scheduleJob(ctx context.Context) error {
+	return c.rescheduleJob(ctx)
+}
+
+func (c *CronDurationJob) updateJob(ctx context.Context) error {
+	return c.rescheduleJob(ctx)
+}
+
+func (c *CronDurationJob) rescheduleJob(ctx context.Context) error {
+	next, err := c.durationToNextStart()
+	if err != nil {
+		return err
+	}
+
+	if err := c.Scheduler.ScheduleJob(
+		quartz.NewJobDetailWithOptions(
+			c,
+			c.key,
+			&quartz.JobDetailOptions{
+				Replace: true,
+			},
+		),
+		quartz.NewRunOnceTrigger(next),
+	); err != nil {
+		return err
+	}
+
+	now := time.Now()
+	status := fleet.ScheduleStatus{
+		Active: false,
+		NextStartTime: metav1.Time{
+			Time: now.In(c.Location).Add(next),
+		},
+		MatchingClusters: c.MatchingClusters,
+	}
+
+	return setScheduleStatus(ctx, c.client, c.Schedule, status)
+}
+
+func (c *CronDurationJob) executeStart(ctx context.Context) error {
+	c.Started = true
+
+	// Recalculate matching clusters at execution time. This ensures that any
+	// changes to cluster labels that occurred since the last reconciliation
+	// are included. The controller's watchers only trigger reconciles for
+	// clusters that are already part of a schedule.
+	clusters, err := matchingClusters(ctx, c.Matcher, c.client, c.Schedule.Namespace)
+	if err != nil {
+		return err
+	}
+
+	// Sets ActiveSchedule to true for all matching clusters.
+	for _, cluster := range clusters {
+		if err := setClusterActiveSchedule(context.Background(), c.client, cluster, c.Schedule.Namespace, true); err != nil {
+			return err
+		}
+	}
```

Comment:
Isn't this redundant with `setClustersScheduled` called below, which calls `setClusterScheduled` on each of those clusters, setting both `cluster.Status.Scheduled` and `cluster.Status.ActiveSchedule`? :thinking:

Reply:
Indeed, and moreover, the call in the highlighted lines is not using the right context. :P 

I've deleted the second block that only updated `Status.Scheduled` and fixed the context in the remaining call.